### PR TITLE
#1488 BUG: Twilio delivered messages getting stuck in sent

### PR DESCRIPTION
**Pull request by EvanParish**
# Description

These changes are in response to Twilio messages getting stuck in "sent" status. These changes should make it less likely that the messages will be stuck when updating statuses. 

issue #1488 

## Type of change

Please check the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

How has this been tested? (e.g. Unit tests, Tested locally, Tested as a GitHub action, Depoyed to dev, etc)  
Please provide relevant information and / or links to demonstrate the functionality or fix. (e.g. screenshots, link to deployment, regression test results, etc)

- [x] unit tests pass locally
- [x] unit tests pass on GitHub
- [x] [Deployed to Dev](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/6671585313)
- [x] [QA suite passes](https://github.com/department-of-veterans-affairs/notification-api-qa/actions/runs/6696595449)

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] The ticket is now moved into the DEV test column

### CODE REVIEW

File: app/celery/process_delivery_status_result_tasks.py [hunk 1]
```diff
@@ -18,20 +19,7 @@
...
-    Notification,
-    NOTIFICATION_PREFERENCES_DECLINED,
-)
-
-FINAL_STATUS_STATES = [
-    NOTIFICATION_DELIVERED,
-    NOTIFICATION_PERMANENT_FAILURE,
-    NOTIFICATION_TECHNICAL_FAILURE,
-    NOTIFICATION_PREFERENCES_DECLINED,
-]
```
Comment (k-macmillan) [inline app/celery/process_delivery_status_result_tasks.py:34]: Weird that flake8 didn't stop this from getting merged the first time

File: app/dao/notifications_dao.py [hunk 2]
```diff
@@ -54,14 +56,21 @@
...
     NOTIFICATION_TEMPORARY_FAILURE
-}
+)
+
+FINAL_STATUS_STATES = (
+    NOTIFICATION_DELIVERED,
+    NOTIFICATION_PERMANENT_FAILURE,
+    NOTIFICATION_TECHNICAL_FAILURE,
+    NOTIFICATION_PREFERENCES_DECLINED,
+)
```
Comment (k-macmillan) [inline app/dao/notifications_dao.py:73]: Import this in `process_pinpoint_receipt_tasks.py` too.

File: app/dao/notifications_dao.py [hunk 3]
```diff
@@ -104,11 +113,88 @@ def country_records_delivery(phone_prefix):
...
 
+def _get_notification_status_update_statement(notification_id: str, incoming_status: str, **kwargs):
+    """
+    Generates an update statement for the given notification id and status.
+    Preserves status order and ensures a transient status will not override a final status.
+    :param notification_id: String value of notification uuid to be updated
+    :param status: String value of the status the given notification will attempt to update to
+    :return: An update statement to be executed, or None
+    """
+    current_status = Notification.query.filter(Notification.id == notification_id).one().status
```
Comment (k-macmillan) [inline app/dao/notifications_dao.py:124]: This does not appear to follow sqlalchemy 2.0 query syntax. 
Should be in the form of `select(Notification).where(Notification.id == notification_id)` etc.

File: app/dao/notifications_dao.py [hunk 4]
```diff
@@ -104,11 +113,88 @@ def country_records_delivery(phone_prefix):
...
+        )
+        return notification
+    except Exception as e:
+        current_app.logger.error(
+            'An error occured when attempting to update notification %s to status %s - The exception %s',
+            notification.id, status, e
+        )
+        return notification
+
+    return Notification.query.filter(Notification.id == notification.id).one()
```
Comment (k-macmillan) [inline app/dao/notifications_dao.py:197]: 2.0 syntax please

File: app/dao/notifications_dao.py [hunk 5]
```diff
@@ -190,19 +251,74 @@ def update_notification_status_by_reference(reference, status):
...
+        )
+        return None
+    except Exception as e:
+        current_app.logger.error(
+            'Unexpected exception thrown attempting to update notification %s, given parameters for updating '
+            'notification may be incorrect - Exception: %s', id, e
+        )
+        return None
+
+    return Notification.query.filter(Notification.id == id).one()
```
Comment (k-macmillan) [inline app/dao/notifications_dao.py:319]: 2.0 syntax please.

**Review by k-macmillan — CHANGES_REQUESTED** [review]

**Review by k-macmillan — APPROVED** [review]

**Review by nikolai-efimov — APPROVED** [review]

**Pull request merged**